crates/hyperdrive-math/src/lp/math.rs

```diff
@@ -281,6 +301,652 @@ impl State {
         Ok(net_flat_trade)
     }
 
+    /// Calculates the amount of withdrawal shares that can be redeemed and
+    /// the share proceeds the withdrawal pool should receive given the
+    /// pool's current idle liquidity. We use the following algorithm to
+    /// ensure that the withdrawal pool receives the correct amount of
+    /// shares to (1) preserve the LP share price and (2) pay out as much
+    /// of the idle liquidity as possible to the withdrawal pool:
+    ///
+    /// 1. If `y_s * t_s <= y_l * t_l` or
+    ///    `y_max_out(I) >= y_s * t_s - y_l * t_l`, set `dz_max = I` and
+    ///    proceed to step (3). Otherwise, proceed to step (2).
+    /// 2. Solve `y_max_out(dz_max) = y_s * t_s - y_l * t_l` for `dz_max`
+    ///    using Newton's method.
+    /// 3. Set `dw = (1 - PV(dz_max) / PV(0)) * l`. If `dw <= w`, then
+    ///    proceed to step (5). Otherwise, set `dw = w` and continue to
+    ///    step (4).
+    /// 4. Solve `PV(0) / l = PV(dz) / (l - dw)` for `dz` using Newton's
+    ///    method if `y_l * t_l != y_s * t_s` or directly otherwise.
+    /// 5. Return `dw` and `dz`.
+    ///
+    ///  Returns (withdrawal_shares_redeemed, share_proceeds, success)
+    pub fn calculate_distribute_excess_idle(
+        &self,
+        current_block_timestamp: U256,
+        active_lp_total_supply: FixedPoint,
+        withdrawal_shares_total_supply: FixedPoint, // withdraw shares - ready to withdraw
+        max_iterations: u64,
+    ) -> Result<(FixedPoint, FixedPoint)> {
+        // Steps 1 and 2: Calculate the maximum amount the share reserves can be
+        // debited. If the effective share reserves or the maximum share
+        // reserves delta can't be calculated or if the maximum share reserves
+        // delta is zero, idle can't be distributed.
+        let success = match self.effective_share_reserves_safe() {
+            Ok(_) => true,
+            // The error is safe from the calculation, panics are not.
+            Err(_) => false,
+        };
+        if success == false {
+            return Ok((fixed!(0), fixed!(0)));
+        }
+        let (max_share_reserves_delta, success) =
+            self.calculate_max_share_reserves_delta(current_block_timestamp)?;
```

Comment:
Should we call this `calculate_max_share_reserves_delta_safe`?

Reply:
yeah it _is_ using a success flag